### PR TITLE
add LabelSelectorSpreadPriority for ReplicaSet in scheduler

### PR DESCRIPTION
**Pull request by mqliang**
replace ReplicationController to ReplicaSet in scheduler

@bgrant0607 @madhusudancs

### CODE REVIEW

**Comment by bgrant0607** [comment]
@mqliang Most users are going to be using ReplicationController for a while, so we can't remove the RC-based functionality from the scheduler yet. We should add ReplicaSet, but keep  ReplicationController.

**Comment by bgrant0607** [comment]
cc @janetkuo @kargakis

**Comment by davidopp** [comment]
+1 to @bgrant0607 's comment. Thanks for noticing this, @mqliang .

**Comment by mqliang** [comment]
I think I got it: We can replace ReplicationController to ReplicaSet in DeploymentController because k8s v1.0 has no DeploymentController. But for `selector_spreading` functionality in scheduler, we must make it back compatible with v1.0. Hence, we should ***ADD*** ReplicaSet, and keep ReplicationController.

**Comment by davidopp** [comment]
Instead of creating a new LabelSelectorSpread type and associated CalculateSpreadPriority(), it would be better to combine these into the existing SelectorSpread and its CalculateSpreadPriority(). So you would have

```
type SelectorSpread struct {
	podLister        algorithm.PodLister
	serviceLister    algorithm.ServiceLister
	controllerLister algorithm.ControllerLister
	replicaSetLister algorithm.ReplicaSetLister
}
```
and in SelectorSpread.CalculateSpreadPriority() you would do

```
	services, err := s.serviceLister.GetPodServices(pod)
	if err == nil {
		for _, service := range services {
			selectors = append(selectors, labels.SelectorFromSet(service.Spec.Selector))
		}
	}
	controllers, err := s.controllerLister.GetPodControllers(pod)
	if err == nil {
		for _, controller := range controllers {
			selectors = append(selectors, labels.SelectorFromSet(controller.Spec.Selector))
		}
	}
	rss, err := s.replicaSetLister.GetPodReplicaSets(pod)
	if err == nil {
		for _, rs := range rss {
			if selector, err := unversioned.LabelSelectorAsSelector(rs.Spec.Selector); err == nil {
				selectors = append(selectors, selector)
			}
		}
	}
```

Once you have created the `selectors` list, there's really no difference between how you handle services, ReplicationControllers, or ReplicaSets. So I don't think there's any need to add a new LabelSelectorSpreadPriority priority function. You can just add ReplicaSet to existing SelectorSpread.

Does that sound reasonable?

**Comment by davidopp** [comment]
ping @mqliang -- I'd really like to get this into 1.2

**Comment by mqliang** [comment]
@davidopp Updated. PTAL.

File: plugin/pkg/scheduler/factory/factory.go [hunk 1]
```diff
@@ -229,6 +234,11 @@ func (f *ConfigFactory) CreateFromKeys(predicateKeys, priorityKeys sets.String,
 	// Cache this locally.
 	cache.NewReflector(f.createControllerLW(), &api.ReplicationController{}, f.ControllerLister.Store, 0).RunUntil(f.StopEverything)
 
+	// Watch and cache all ReplicaSet objects. Scheduler needs to find all pods
+	// created by the same services or ReplicaSets, so that it can spread them correctly.
```
Comment (davidopp) [inline plugin/pkg/scheduler/factory/factory.go]: s/services or ReplicaSets/services or ReplicationControllers or ReplicaSets/

also, please make the same fix on L233 and L228

**Comment by davidopp** [comment]
This looks great -- thanks! I just had one very minor comment.

**Comment by mqliang** [comment]
@davidopp Comments addressed. PTAL

**Comment by davidopp** [comment]
LGTM

Thanks!

**Pull request merged**